src/services/kratos/index.ts

```diff
@@ -0,0 +1,194 @@
+  const method = "password"
+
+  let result: AxiosResponse<SuccessfulSelfServiceLoginWithoutBrowser>
+
+  try {
+    result = await kratosPublic.submitSelfServiceLoginFlow(flow.data.id, {
+      identifier,
+      method,
+      password,
+    })
+  } catch (err) {
+    if (err.message === "Request failed with status code 400") {
+      return new LikelyNoUserWithThisPhoneExistError(err)
+    }
+
+    if (err.message === "Request failed with status code 401") {
+      return new AuthenticationKratosError(err)
+    }
+
+    return new UnknownKratosError(err)
+  }
+
+  const sessionToken = result.data.session_token as KratosSessionToken
+  const kratosUserId = result.data.session.identity.id as KratosUserId
+  return { sessionToken, kratosUserId }
+}
+
+export const createKratosUserForPhoneNoPasswordSchema = async (
+  phone: PhoneNumber,
+): Promise<CreateKratosUserForPhoneNoPasswordSchemaResponse | KratosError> => {
+  const flow = await kratosPublic.initializeSelfServiceRegistrationFlowWithoutBrowser()
+
+  const traits = { phone }
+  const method = "password"
+
+  let result: AxiosResponse<SuccessfulSelfServiceRegistrationWithoutBrowser>
+
+  try {
+    result = await kratosPublic.submitSelfServiceRegistrationFlow(flow.data.id, {
+      traits,
+      method,
+      password,
+    })
+  } catch (err) {
+    if (err.message === "Request failed with status code 400") {
+      return new LikelyUserAlreadyExistError(err)
+    }
+
+    return new UnknownKratosError(err)
+  }
+
+  const sessionToken = result.data.session_token as KratosSessionToken
+  const kratosUserId = result.data.identity.id as KratosUserId
+
+  return { sessionToken, kratosUserId }
+}
+
+export const upgradeUserToPasswordSchema = async ({
+  kratosUserId,
+  password,
+}: {
+  kratosUserId: KratosUserId
+  password: string
+}) => {
+  const { data: identity } = await kratosAdmin.adminGetIdentity(kratosUserId)
+
+  if (identity.schema_id !== "phone_v0") {
+    return new IncompatibleSchemaUpgradeError()
+  }
+
+  const adminIdentity: AdminUpdateIdentityBody = {
+    ...identity,
+    credentials: { password: { config: { password } } },
+    state: "active",
+    schema_id: "phone_with_password_v0",
+  }
+
+  const { data: newIdentity } = await kratosAdmin.adminUpdateIdentity(
+    kratosUserId,
+    adminIdentity,
+  )
+  return newIdentity
+}
+
+export const validateKratosToken = async (
+  KratosSessionToken: KratosSessionToken,
+): Promise<KratosUserId | KratosError> => {
+  let session: Session
+
+  try {
+    const { data } = await kratosPublic.toSession(KratosSessionToken)
+    session = data
+  } catch (err) {
+    if (err.message === "Request failed with status code 401") {
+      return new AuthenticationKratosError(err)
+    }
+    return new UnknownKratosError(err)
+  }
+
+  // TODO active: true
+
+  return session.identity.id as KratosUserId
+}
+
+export const activateUser = async (
+  kratosUserId: KratosUserId,
+): Promise<void | KratosError> => {
```

Comment:
I think we have just been returning `true | XXX` in situations like this.

Reply:
We discussed exactly this pattern and decided on `true | ` - but we can bring it up again.

Reply:
there are a lot of `<void | Error>` as well

<img width="512" alt="image" src="https://user-images.githubusercontent.com/11156370/193458888-89b258cc-69dc-48df-93d7-8768069cf47d.png">